### PR TITLE
feat: add Google open source scale bar

### PR DESCRIPTION
**Pull request by sargunv**
closes https://github.com/sargunv/maplibre-compose/issues/157

CleanShot 2024-12-20 at 15 04 14@2x

### CODE REVIEW

**Comment by sargunv** [comment]
* **#159** <a href="https://app.graphite.dev/github/pr/sargunv/maplibre-compose/159?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/sargunv/maplibre-compose/159?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by westnordost — COMMENTED** [review]
Ha, cool, where did you find **that** implementation?

Anyway, a few notes after reading its source:

## 1. units

Users using a non-US-locale definitely don't want to see feet and miles, it is just useless extra noise for them. At the least, there should be an option with reasonable defaults.

Weird, I thought I did some research and posted it here somewhere on which locales imperial and on which locales SI units would be expected. Anyway, did the research again. 
Reasonable defaults could be:


- <a href="https://ent8r.github.io/blacklistr/?data=US">Show imperial only<br></a>
- <a href="https://ent8r.github.io/blacklistr/?data=// slow transition to metric%0ABZ,LR,MM,%0A// US dependencies (mainly imperial, some metric in use)%0AAS,GU,MH,MP,PR,PW,VI,WS%0A// GB and British dependencies (mainly metric, some imperial remains)%0AGB,AG,AI,BM,BS,FK,GD,GG,GI,GS,IM,JE,KN,KY,LC,MS,TC,VG,VC,">Show imperial and meters<br></a>
- Show only meters: all others

Source: https://github.com/streetcomplete/countrymetadata/blob/master/data/lengthUnits.yml (result of some research I did years ago on what's displayed on street signs)

The list of "metric + imperial" is very lavish, officially, most of them are actually metric only. But well, [officially, the US is also transitioning to metric units](https://www.nist.gov/pml/owm/metric-si), so..., anyway, maybe only have two categories: imperial+metric and metric only.

## 2. Unit localization

This is probably not necessary. At least the abbreviations should be the same everywhere. "m" is meter everywhere, "k" is kilo everywhere, and regarding "ft" and "mi", they are only used in English-speaking countries anyway, except Puerto Rico. Looking at the [Spanish wikipedia article](https://es.wikipedia.org/wiki/Pie_(unidad)) for Foot, there doesn't seem to be a Spanish abbreviation, maybe because "pie(s)" is short enough or they also use "ft"?

## 3. line "shadows"

For the lines and the text, normal (and fake) shadows are used. On a busy map, this is not well visible. I propose to do it the same as with other labels on the map: Have an **outline**, aka **halo**.
(How to do this with text?: Draw the text twice, but one time with 
```kotlin
val stroke = Stroke(haloWidth.toPx() * 2, cap = StrokeCap.Round, join = StrokeJoin.Round)
val style = localTextStyle.copy(drawStyle = stroke)
```
It would be nice if the halo width and also the line width of the scale bar could be controlled via parameters.

File: demo-app/src/commonMain/kotlin/dev/sargunv/maplibrecompose/demoapp/demos/EdgeToEdgeDemo.kt [hunk 1]
```diff
@@ -42,16 +43,12 @@ object EdgeToEdgeDemo : Demo {
           cameraState = cameraState,
           styleState = styleState,
           ornamentSettings =
-            OrnamentSettings(
-              padding = padding,
-              isCompassEnabled = false,
-              isLogoEnabled = false,
-              isAttributionEnabled = false,
-            ),
+            OrnamentSettings.AllDisabled.copy(isLogoEnabled = true, padding = padding),
```
[the remark this answers is not in the document]
Comment (sargunv) [reply demo-app/src/commonMain/kotlin/dev/sargunv/maplibrecompose/demoapp/demos/EdgeToEdgeDemo.kt:46]: Overlaying Compose UI on top of interop views is unreliable on Desktop and not available on Web. 

Maybe solvable, but I want to maintain support for the upstream MapLibre ornaments at least until we've got support for those platforms to try the Compose-UI overlays. 

There's also the question of whether the app is even using Material style at all (my app, I plan to use SwiftUI on iOS and Material on Android, with Compose driving the map on both). So having Material controls being opt-in makes sense to me.

**Comment by sargunv** [comment]
Re: units

In the US, Google Maps shows both units on the scale bar but defaults to feet/miles for navigation. Personally I appreciate seeing both because I generally think in terms of km, communicate IRL in terms of mi, and communicate online in terms of km.

I think it'd be reasonable to default to both in non-SI or transitioning regions and just SI everywhere else. There can be an option for ft/mi only display but I wouldn't make it the default for the scale bar anywhere.

Curious, in SI regions, does Google not show both units on the scale bar?

Re: localizing abbreviations

Definitely something I gotta research more. 

Re: shadows

Seems reasonable, though looking at the style switcher demo it seems to be reasonably visible already on all the styles in New York City.

**Comment by sargunv** [comment]
The current implementation was borrowed from https://github.com/googlemaps/android-maps-compose/blob/main/maps-compose-widgets%2Fsrc%2Fmain%2Fjava%2Fcom%2Fgoogle%2Fmaps%2Fandroid%2Fcompose%2Fwidgets%2FScaleBar.kt

File: demo-app/src/commonMain/kotlin/dev/sargunv/maplibrecompose/demoapp/demos/EdgeToEdgeDemo.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (westnordost) [reply demo-app/src/commonMain/kotlin/dev/sargunv/maplibrecompose/demoapp/demos/EdgeToEdgeDemo.kt:46]: Whoa? What? How, why? Why would it not work to put compose UI content on top of non-UI content? That sounds like a pretty huge constraint.

**Comment by westnordost** [comment]
> Curious, in SI regions, does Google not show both units on the scale bar?

No, the app shows both, regardless of locale. The website shows only meters.

But I wouldn't look at Google Maps for good map design anyway, it got worse and worth over the last few years.
(Apparently, the most important things in my neighborhood, i.e. 2x2km² area are a certain ambulatory care, a cooking school and a Subway, the only POIs that are shown when zoomed out quite a lot. Not the metro stations, not the parks or iconic buildings such as churches)

**Comment by westnordost** [comment]
> shadows

Looking at google maps app, it looks like they already implemented exactly that. (A "halo" around the line and text.) So, the open source scalebar deviates from what is shown in the app, it seems.

**Comment by sargunv** [comment]
The more I look at the Google Maps scale bar, the weirder it gets. 

It doesn't tell you the scale at the center "target" of the camera like ours, nor at the location of the scale bar itself like their open source one. On my phone at least, it shows the scale of the point about 1/3 down from the top of the screen, far from the bar at the bottom. 

Or in this screenshot, the scale is for where the equator is:

**Comment by sargunv** [comment]
Re: localizing unit symbols 

I think in Spanish feet is just pies or '. There doesn't seem to be an analogue to ft. Inches aren't relevant to us for maps but pulgadas in Spanish is shortened to pulg or ". 

SI symbols seem to be pretty universal in the latin script, but some other writing systems have their own symbols, for example in Chinese, meter is 米 and kilometer is 公里. In Arabic, they're م and كم. 

See the Baidu Maps scale bar:

**Comment by westnordost** [comment]
wow, cool. I like that character, looks like a star

File: demo-app/src/commonMain/kotlin/dev/sargunv/maplibrecompose/demoapp/demos/EdgeToEdgeDemo.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (sargunv) [reply demo-app/src/commonMain/kotlin/dev/sargunv/maplibrecompose/demoapp/demos/EdgeToEdgeDemo.kt:46]: This is the issue for desktop: https://youtrack.jetbrains.com/issue/CMP-2926. There's a partial fix that requires opting in to a feature flag, and only works on some desktop platforms: https://github.com/JetBrains/compose-multiplatform-core/pull/915. Before that fix (and even today by default), interop swing panels are always displayed on top of Compose UI. 

On web, there is no interop html element available yet. There are some examples floating around on GitHub (https://github.com/Hamamas/Kotlin-Wasm-Html-Interop) but they put the element on top of the Compose UI canvas. Maybe it's possible to "punch a hole" in the canvas to enable layering between MapLibre and Compose UI, but nobody has done it yet to my knowledge.

**Pull request merged**